From d4258e9a4b0a294f25ab0d26eac0f0fa7725df12 Mon Sep 17 00:00:00 2001
From: Pedro Palazon Candel <pedro@joyent.com>
Date: Tue, 4 Jun 2019 18:13:12 +0200
Subject: [PATCH] TRITON-1708 CloudAPI should validate package.brand value
 against image.type before setting machines brand Reviewed by: Todd Whiteman
 <todd.whiteman@joyent.com> Approved by: Todd Whiteman
 <todd.whiteman@joyent.com>

---
 lib/machines.js               | 16 ++++++++++++++
 test/getCreateOptions.test.js | 39 ++++++++++++++++++++++++++++++++++-
 2 files changed, 54 insertions(+), 1 deletion(-)

diff --git a/lib/machines.js b/lib/machines.js
index eda4ffa..da907e9 100644
--- a/lib/machines.js
+++ b/lib/machines.js
@@ -57,6 +57,7 @@ var PKG_USED_PARAMS = ['uuid', 'max_physical_memory', 'name', 'version',
 var DEFAULT_CONTAINER_BRAND = 'joyent';
 var DEFAULT_HVM_BRAND = 'kvm';
 var VALID_BRANDS = ['bhyve', 'joyent', 'joyent-minimal', 'kvm', 'lx'];
+var VALID_HVM_BRANDS = ['bhyve', 'kvm'];
 
 var sprintf = util.format;
 
@@ -415,6 +416,21 @@ function getCreateOptions(req) {
                 'Package requires unknown brand "' +
                 safeBrandName(pkg.brand) + '"');
         }
+        if (VALID_HVM_BRANDS.indexOf(pkg.brand) !== -1) {
+            if (!img.type || img.type !== 'zvol') {
+                throw new InvalidArgumentError(
+                    'Package requires brand "' +
+                    safeBrandName(pkg.brand) +
+                    '" and image.type is not "zvol"');
+            }
+        } else if (VALID_HVM_BRANDS.indexOf(pkg.brand) === -1) {
+            if (img.type === 'zvol') {
+                throw new InvalidArgumentError(
+                    'Package requires brand "' +
+                    safeBrandName(pkg.brand) +
+                    '" and image.type is incompatible ("zvol")');
+            }
+        }
         brand = pkg.brand;
     } else if (img.type === 'zvol') {
         brand = DEFAULT_HVM_BRAND;
diff --git a/test/getCreateOptions.test.js b/test/getCreateOptions.test.js
index 372d7b3..f3d6c16 100644
--- a/test/getCreateOptions.test.js
+++ b/test/getCreateOptions.test.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright 2019 Joyent, Inc.
  */
 
 var fs = require('fs');
@@ -313,6 +313,43 @@ test('getCreateOptions blows up when pkg.brand and img.requirements.brand ' +
     t.end();
 });
 
+test('getCreateOptions blows up when pkg.brand and img.type conflict',
+    function (t) {
+    var createOpts;
+    var req, req2;
+
+    req = buildReq({
+        img: IMAGES['ubuntu-bhyve-17.10-noBrandReq'],
+        pkg: {
+            brand: 'lx'
+        }
+    });
+
+    t.throws(function _getPayload() {
+        createOpts = getCreateOptions(req);
+        t.equal(createOpts, undefined, 'should not have createOpts');
+    // JSSTYLED
+    }, /Package requires brand "lx" and image.type is incompatible \("zvol"\)/,
+        'conflicting pkg.brand and img.type should result in exception');
+
+
+    req2 = buildReq({
+        img: IMAGES['smartos-1.6.3'],
+        pkg: {
+            brand: 'bhyve'
+        }
+    });
+
+    t.throws(function _getPayload2() {
+        createOpts = getCreateOptions(req2);
+        t.equal(createOpts, undefined, 'should not have createOpts');
+    // JSSTYLED
+    }, /Package requires brand "bhyve" and image.type is not "zvol"/,
+        'conflicting pkg.brand and img.type should result in exception');
+
+    t.end();
+});
+
 
 test('test safeBrandName', function test_safe_brand_name(t) {
     // undefined should blow the assert.string
-- 
2.21.0

